dev/tools/localization/gen_l10n_types.dart

```diff
@@ -374,19 +374,37 @@ class AppResourceBundle {
     }
 
     String localeString = resources['@@locale'] as String;
+
+    // If the locale string was not defined in @@locale, look for the first
+    // instance of an ISO 639-1 language code, matching exactly.
     if (localeString == null) {
-      final RegExp filenameRE = RegExp(r'^[^_]*_(\w+)\.arb$');
-      final RegExpMatch match = filenameRE.firstMatch(file.path);
-      localeString = match == null ? null : match[1];
+      final String fileName = file.path
+        // Clean out the path to only contain the filename.
+        .split('/').last
+        // Remove the file type suffix.
+        .split('.')[0];
+
+      // Separate out the filename by the underscores in the string.
+      final List<String> fileNameSeparatedByUnderscores = fileName.split('_');
+
+      for (int index = 0; index < fileNameSeparatedByUnderscores.length; index += 1) {
```

Comment:
This a nice improvement over what we have. The generalization of the idea might be a little better: match filenames with a suffix that matches a legitimate locale string preceded by an underscore. You can use https://pub.dev/documentation/intl/latest/locale/Locale/tryParse.html to check each possible suffix.